--- a/misc/terminfo.src	2020-01-19 02:09:38.000000000 +0100
+++ b/misc/terminfo.src	2021-05-25 16:07:16.093288876 +0200
@@ -924,6 +924,49 @@
 	smul=\E[4m, u6=\E[%i%p1%d;%p2%dR, u7=\E[6n,
 	vpa=\E[%i%p1%dd,
 
+#### Cobalt
+
+cobalt-16color|Cobalt terminal with 16 colors,
+	am, AX, bce, msgr, npc, xenl,
+	colors#16, pairs#256, op=\E[39;49m,
+	setab=\E[%?%p1%{8}%<%t4%p1%d%e10%p1%{8}%-%d%;m,
+	setaf=\E[%?%p1%{8}%<%t3%p1%d%e9%p1%{8}%-%d%;m,
+	ht=^I, it#8,
+	rc=\E8, sc=\E7,
+	el1=\E[1K, ind=\n, ri=\E[T,
+	bold=\E[1m, rev=\E[7m, smso=\E[7m, rmso=\E[27m,
+	sgr0=\E[m, sgr=\E[0%?%p6%t;1%;%?%p1%p3%|%t;7%;m,
+	nel=\n, rs1=\Ec, civis=\E[?25l, cnorm=\E[?25h,
+	smcup=\E[?1049h, rmcup=\E[?1049l,
+	kdch1=\E[3~, kend=\E[F, kf1=\E[OP, kf2=\E[OQ, kf3=\E[OR, kf4=\E[OS,
+	kf5=\E[15~, kf6=\E[17~, kf7=\E[18~, kf8=\E[19~, kf9=\E[20~, kf10=\E[21~,
+	kf11=\E[23~, kf12=\E[24~, kich1=\E[2~, knp=\E[6~, kpp=\E[5~, kent=\n,
+	use=ansi+local, use=ansi+erase, use=ansi+rca, use=ansi+cup,
+	use=ansi+arrows, use=ecma+index,
+
+# The Cobalt terminal (except for -16color) always supports 16 million colors.
+# Unfortunately ncurses support for this is far from perfect. Palette colors
+# and RGB colors share a common namespace. When RGB support is advertised only
+# 8 palette colors are usable but applications might expect 16 palette colors
+# instead. Also these 8 palette colors mean that 8 RGB colors are missing.
+# Really ncurses should add a new capability to select RGB colors without
+# having to lose the palette colors.
+# Therefore the cobalt entry only uses the 256 palette colors. If you want to
+# use the RGB colors you can set TERM to cobalt-direct.
+cobalt|Cobalt terminal,
+	colors#256, pairs#32767,
+	setab=\E[48;5;%p1%dm,
+	setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
+	use=cobalt-16color,
+
+cobalt-direct|Cobalt terminal with direct colors,
+	colors#0x1000000, RGB, pairs#0x10000,
+	setab=\E[%?%p1%{8}%<%t4%p1%d%e48;2;%p1%{65536}%/%d;%p1%{256}
+	      %/%{255}%&%d;%p1%{255}%&%d%;m,
+	setaf=\E[%?%p1%{8}%<%t3%p1%d%e38;2;%p1%{65536}%/%d;%p1%{256}
+	      %/%{255}%&%d;%p1%{255}%&%d%;m,
+	use=cobalt-16color,
+
 #### Linux consoles
 #
 
